src/errors.ts

```diff
@@ -1,5 +1,31 @@
 import * as z from 'zod'
-import type { ErrorWithMessage, SchemaError } from './types'
+import type { ErrorData, ErrorWithMessage, SchemaError } from './types'
+
+function parseError(error: unknown): ErrorData {
+  if (isErrorWithErrorData(error)) {
+    return {
+      errors: error.errors || [],
+      inputErrors: error.inputErrors || [],
+      environmentErrors: error.environmentErrors || [],
+    }
+  }
+
+  return {
+    errors: [toErrorWithMessage(error)],
+    inputErrors: [],
+    environmentErrors: [],
+  }
+}
+
+function isErrorWithErrorData(error: unknown): error is Partial<ErrorData> {
```

Comment:
Instead of typing this as a Partial I would rather have concrete subtypes. This would make the types easier to read and reason about. For instance, today `ErrorData` is defined as:
```ts
type ErrorData = Omit<ErrorResult, 'success'>
```

But we could build up from simpler error types until we get to error result:

```ts
type ErrorData = GeneralErrors & InputErrors & EnvironmentErrors
type ErrorResult = { success: false } & ErrorData
```

Reply:
In this case you cloud have custom exceptions like the one suggested by @gustavoguichard but more than one to avoid having to check the presence of fields in runtime. It's best to ensure the shape of the error during type check and reduce our runtime logic.